src/sentry/monitors/tasks.py

```diff
@@ -392,3 +411,58 @@ def mark_checkin_timeout(checkin_id: int, ts: datetime, **kwargs):
         )
 
         mark_failed(checkin, ts=most_recent_expected_ts)
+
+
+@instrumented_task(
+    name="sentry.monitors.tasks.detect_broken_monitor_envs",
+    max_retries=0,
+    time_limit=15 * 60,
+    soft_time_limit=10 * 60,
+    record_timing=True,
+)
+def detect_broken_monitor_envs():
+    current_time = django_timezone.now()
+    open_incidents_qs = MonitorIncident.objects.select_related("monitor").filter(
+        resolving_checkin=None,
+        starting_timestamp__lte=(current_time - timedelta(days=NUM_DAYS_BROKEN_PERIOD)),
+        # TODO(davidenwang): When we want to email users, remove this filter
+        monitorenvbrokendetection__isnull=True,
+    )
+    for open_incidents in chunked(
+        RangeQuerySetWrapper(
+            open_incidents_qs,
+            step=1000,
+        ),
+        1000,
+    ):
+        for open_incident in open_incidents:
+            try:
+                organization = Organization.objects.get_from_cache(
+                    id=open_incident.monitor.organization_id
+                )
+                if not features.has(
+                    "organizations:crons-broken-monitor-detection", organization=organization
+                ):
+                    continue
+            except Organization.DoesNotExist:
+                continue
+
+            # verify that the most recent check-ins have been failing
+            recent_checkins = (
+                MonitorCheckIn.objects.filter(monitor_environment=open_incident.monitor_environment)
+                .order_by("-date_added")
+                .values("status")[:NUM_CONSECUTIVE_BROKEN_CHECKINS]
+            )
+            if len(recent_checkins) != NUM_CONSECUTIVE_BROKEN_CHECKINS or not all(
+                checkin["status"]
+                in [CheckInStatus.ERROR, CheckInStatus.TIMEOUT, CheckInStatus.MISSED]
+                for checkin in recent_checkins
+            ):
+                continue
+
+            detection, _ = MonitorEnvBrokenDetection.objects.get_or_create(
+                monitor_incident=open_incident, defaults={"detection_timestamp": current_time}
+            )
+            if not detection.user_notified_timestamp:
+                # TODO(davidenwang): This is where we would implement email sending logic
+                pass
```

Comment:
Wasn't the idea that we were going to try and batch these somehow?

Reply:
yea I do need to think more about the appropriate way to do batching here, which could potentially change the logic for running through the list of open incidents. For now though I want to get the dry detection code out and continue to iterate on the task for doing batch emails